airflow/models/taskinstance.py

```diff
@@ -1287,7 +1287,7 @@ class TaskInstance(Base, LoggingMixin):
     queued_dttm = Column(UtcDateTime)
     queued_by_job_id = Column(Integer)
     pid = Column(Integer)
-    executor_config = Column(ExecutorConfigType(pickler=dill))
+    executor_config = Column(ExecutorConfigType(pickler=cloudpickle))
```

Comment:
Hmmm. That changes things a bit.  This means that we really need either dil or cloudpickle not only as optional dependency - but also as required one - becuase basically we have to be able to always store executor_config. This is a bit of a bummer, I am afraid  - it means that we cannot really make dil an optional dependency eventually

Reply:
So, as i can see now we want to replace dill with cloudpickle more then just add both as optional to use both, but we don't have right now a safe way to replace one with another? :)

Reply:
@eladkal
What do you think about the changes here?

Reply:
Maybe we can use in this case here another package, like pickle? I think this is not an optional dependency, so maybe here, until we can fully replace dill with cloudpickle, we can use pickle. The same way we are using it in places where user did not specify use_cloudpickle or use_dill

Reply:
I believe - we cannot use pickle. There are certain executors (K8S executor is the main culprit) that can use configuration that is not picklable because it can use V1Pod and similar classes for custom pod template. 

See the example here where you provide custom Pod defintion via executor-config passed to the task. Such config is serialized when the DAG is parsed and stored in the DB, and deserialized by the executor when it attempts to execute it:

https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/kubernetes_executor.html#pod-override

It's likely however that we could swap the default for `cloudpickle` (or maybe even choose whichever pickling class is available and importable - we could detect it, but then there is an issue for conversion for past task in case of changeover - as @hussein-awala mentioned and implemented for encrypted kwargs: https://github.com/apache/airflow/pull/38358

I still think it's possibe to have the libraries as optional for core (and pre-install both in airflow reference image), but It needs a bit of deliberation and right errors/messaging when there are already task in the DB which have been serialized using one of them, but when the library is missing - basically handling the case where someone had tasks created in the DB using `dill` and starts airflow without `dill` - we should raise an error and tell the user to install `dill` to make the on-the-flight conversion of old values).

Reply:
@potiuk ,
If we can pre-install both `dill` and `cloudpickle` packages, I think we could implement an optional feature as follows:

1. Specify option for the package somewhere and describe it in the documentation. Default value is `cloudpickle`.
2. Add migration that reformats models (`TaskInstance`, `DagPickle`) so they would also store information about the package used for serialization (in our case - dill==3.1.1.1). For instance, the new format could be:
```python
{
  "pickler": {"name": "dill", "version": "3.1.1"},
  "object": <object>
}
 ```
  After the migration is complete, we always know which package serialized the value. This information will help us perform the migration rollback properly, and reformatting existing objects whenever users switch from one package to another. 

3. Integrate into the Airflow startup process a check for the serialization package selected, and run the script that migrates saved objects from one pickler package to another.
4. Adjust the `ExecutorConfigType` in order to support the new storage format.

WDYT?

Reply:
Yes, this solution seems way simpler! Thanks for the suggestion. We will update this PR.

Reply:
@potiuk , hi.
I looked into the [serde serialization](https://github.com/apache/airflow/blob/main/airflow/serialization/serde.py#L87), and if it was a right place to look at (please correct me if it wasn't), then I'm afraid its not a robust option. Unlike `dill` and `cloudpickle`, serde's `serialize()` method loses an object's type information similar to what `json.dumps()` does:
```python
from airflow.serialization.serde import deserialize, serialize

data = {1: 1}
serialized = serialize(data)  # {'1': 1}
deserialized = deserialize(serialized)  # {'1': 1}
data == deserialized  # False
```
Because we can't be sure that there are no more corner cases left, I think it's better to rely on some package supported by a dedicated team (`cloudpickle`). However, if I'm overcautious, please tell me that this isn't a critical issue for `TaskInstance.executor_config` and `DagPickle.pickle`.